storage-proofs/core/src/compound_proof.rs

```diff
@@ -70,25 +73,81 @@ where
         groth_params: &'b groth16::MappedParameters<Bls12>,
     ) -> Result<MultiProof<'b>> {
+        let vanilla_proofs = Self::prove_vanilla(pub_params, pub_in, priv_in)?;
+
+        Self::prove_circuit(pub_params, pub_in, vanilla_proofs, groth_params)
+    }
+
+    /// prove_vanilla and prove_circuit combined is equivalent to ProofScheme::prove.
+    fn prove_vanilla(
+        pub_params: &PublicParams<'a, S>,
+        pub_in: &S::PublicInputs,
+        priv_in: &S::PrivateInputs,
+    ) -> Result<Vec<S::Proof>> {
         let partition_count = Self::partition_count(pub_params);
 
         // This will always run at least once, since there cannot be zero partitions.
         ensure!(partition_count > 0, "There must be partitions");
 
-        info!("vanilla_proof:start");
+        info!("prove_vanilla:start");
         let vanilla_proofs = S::prove_all_partitions(
             &pub_params.vanilla_params,
             &pub_in,
             priv_in,
             partition_count,
         )?;
 
-        info!("vanilla_proof:finish");
+        info!("prove_vanilla:finish");
+
+        let sanity_check =
+            S::verify_all_partitions(&pub_params.vanilla_params, &pub_in, &vanilla_proofs)?;
+        ensure!(sanity_check, "sanity check failed");
+
+        Ok(vanilla_proofs)
+    }
+
+    /// prove_vanilla and prove_circuit combined is equivalent to ProofScheme::prove.
+    fn prove_vanilla_with_challenges(
+        pub_params: &PublicParams<'a, S>,
+        pub_in: &S::PublicInputs,
+        priv_in: &S::PrivateInputs,
+        sector_challenges: &BTreeMap<SectorId, Vec<u64>>,
+    ) -> Result<Vec<S::Proof>> {
+        let partition_count = Self::partition_count(pub_params);
+
+        // This will always run at least once, since there cannot be zero partitions.
+        ensure!(partition_count > 0, "There must be partitions");
+
+        info!("prove_vanilla_with_challenges:start");
+        let vanilla_proofs = S::prove_all_partitions_with_challenges(
+            &pub_params.vanilla_params,
+            &pub_in,
+            priv_in,
+            partition_count,
+            sector_challenges,
+        )?;
+
+        info!("prove_vanilla_with_challenges:finish");
 
         let sanity_check =
             S::verify_all_partitions(&pub_params.vanilla_params, &pub_in, &vanilla_proofs)?;
         ensure!(sanity_check, "sanity check failed");
 
+        Ok(vanilla_proofs)
+    }
+
+    /// prove_vanilla and prove_circuit combined is equivalent to ProofScheme::prove.
+    fn prove_circuit<'b>(
+        pub_params: &PublicParams<'a, S>,
+        pub_in: &S::PublicInputs,
+        vanilla_proofs: Vec<S::Proof>,
+        groth_params: &'b groth16::MappedParameters<Bls12>,
+    ) -> Result<MultiProof<'b>> {
+        let partition_count = Self::partition_count(pub_params);
+
+        // This will always run at least once, since there cannot be zero partitions.
+        ensure!(partition_count > 0, "There must be partitions");
+
```

Comment:
We should probably include a sanity check so we never bother trying to create circuit proofs for invalid vanilla proofs.